### PR TITLE
Implement change feed for blocked data sources for AB#15397.

### PR DESCRIPTION
**Pull request by BrianMaki**
# Implements [AB#15397](https://dev.azure.com/qslvic/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/15397)

## Description

- Implemented change feed functionality for block access
- Renamed BlockAccess to BlockAccessAsync in PatientRepository
- Removed ConfigureAwait(true) as it is no longer required in touched. code and areas around touched code 
- Updated unit tests to include change feed functionality
- Added integration test for Block Access and change feed

**Unit and Integration Tests:**

Screenshot 2023-10-25 at 6 23 55 PM

**Functional Tests:**

Screenshot 2023-10-25 at 5 10 56 PM

## Testing

- [x] Unit Tests Updated
- [ ] Functional Tests Updated
- [ ] Not Required

## UI Changes

None

## Items to Review:

-   [General PR Guidelines](https://github.com/bcgov/healthgateway/wiki/PRguidance)

### CODE REVIEW

**Review by BronzBierd — DISMISSED** [review]

File: Apps/AccountDataAccess/src/Patient/PatientRepository.cs [hunk 1]
```diff
@@ -36,8 +38,11 @@ namespace HealthGateway.AccountDataAccess.Patient
     internal class PatientRepository : IPatientRepository
     {
         private const string BlockedAccessKey = "BlockedAccess";
+        private const string BlockedDataSourceChangeFeedKey = "BlockedDataSources";
```
[the remark this answers is not in the document]
Comment (BronzBierd) [reply Apps/AccountDataAccess/src/Patient/PatientRepository.cs:41]: https://dev.azure.com/qslvic/Health%20Gateway/_workitems/edit/16372

**Review by MikeLyttle — APPROVED** [review]

File: Apps/AccountDataAccess/src/Patient/PatientRepository.cs [hunk 2]
```diff
@@ -120,16 +130,23 @@ public async Task BlockAccess(BlockAccessCommand command, CancellationToken ct =
...
             }
             else
             {
-                await this.blockedAccessDelegate.DeleteBlockedAccessAsync(blockedAccess, agentAudit);
+                await this.blockedAccessDelegate.DeleteBlockedAccessAsync(blockedAccess, agentAudit, !this.changeFeedEnabled);
+            }
+
+            if (this.changeFeedEnabled)
+            {
+                await this.messageSender.SendAsync(new[] { new MessageEnvelope(new DataSourcesBlockedEvent(command.Hdid, blockedAccess.DataSources), command.Hdid) }, ct);
```
Comment (ytqsl) [inline Apps/AccountDataAccess/src/Patient/PatientRepository.cs:149]: I'd recommend not reusing DataSource enum in the event, so either use a string in the event or replicate the enum.

File: Apps/Admin/Server/Services/SupportService.cs [hunk 3]
```diff
@@ -208,7 +208,7 @@ public async Task<IList<PatientSupportResult>> GetPatientsAsync(PatientQueryType
         public async Task BlockAccessAsync(string hdid, IEnumerable<DataSource> dataSources, string reason, CancellationToken ct = default)
         {
             BlockAccessCommand blockAccessCommand = new(hdid, dataSources, reason);
-            await this.patientRepository.BlockAccess(blockAccessCommand, ct).ConfigureAwait(true);
+            await this.patientRepository.BlockAccessAsync(blockAccessCommand, ct).ConfigureAwait(true);
```
Comment (ytqsl) [inline Apps/Admin/Server/Services/SupportService.cs:211]: would be nice to remove configureawait too

File: Apps/Common/src/Models/Events/DataSourcesBlockedEvent.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+using System.Collections.Generic;
+using HealthGateway.Common.Data.Constants;
+using HealthGateway.Common.Messaging;
+
+/// <summary>
+/// Represents a data source(s) blocked event
+/// </summary>
+/// <param name="Hdid">The hdid associated with the blocked data source(s)</param>
+/// <param name="DataSources">The data sources to block.</param>
+public record DataSourcesBlockedEvent(string Hdid, IEnumerable<DataSource> DataSources) : MessageBase;
```
Comment (ytqsl) [inline Apps/Common/src/Models/Events/DataSourcesBlockedEvent.cs:27]: Same as above - I'd recommend considering not to share the DataSource enum in the event signature

**Review by ytqsl — APPROVED** [review]

File: Apps/Admin/Server/Services/SupportService.cs (hunk 3 above)
Comment (BrianMaki) [reply Apps/Admin/Server/Services/SupportService.cs:211]: Changed

File: Apps/AccountDataAccess/src/Patient/PatientRepository.cs (hunk 2 above)
Comment (BrianMaki) [reply Apps/AccountDataAccess/src/Patient/PatientRepository.cs]: Changed.

File: Apps/Common/src/Models/Events/DataSourcesBlockedEvent.cs (hunk 4 above)
Comment (BrianMaki) [reply Apps/Common/src/Models/Events/DataSourcesBlockedEvent.cs]: Changed.

File: Apps/Admin/Server/appsettings.json [hunk 5]
```diff
@@ -139,5 +139,13 @@
     "Delegation": {
         "MaxDependentAge": 12,
         "MinDelegateAge": 12
+    },
```
Comment (ytqsl) [inline Apps/Admin/Server/appsettings.json:142]: I think this is redundant, if the settings default to false and not specified, it would still be false

**Review by ytqsl — APPROVED** [review]

File: Apps/Admin/Server/appsettings.json (hunk 5 above)
Comment (BrianMaki) [reply Apps/Admin/Server/appsettings.json:142]: Can discuss again with @sslaws

File: Apps/AccountDataAccess/src/Patient/PatientRepository.cs [hunk 6]
```diff
@@ -146,7 +146,8 @@ public async Task BlockAccessAsync(BlockAccessCommand command, CancellationToken
 
             if (this.changeFeedEnabled)
             {
-                await this.messageSender.SendAsync(new[] { new MessageEnvelope(new DataSourcesBlockedEvent(command.Hdid, blockedAccess.DataSources), command.Hdid) }, ct);
+                IEnumerable<string> dataSourceValues = blockedAccess.DataSources.Select(ds => Enum.GetName(typeof(DataSource), ds))!;
```
Comment (MikeLyttle) [inline Apps/AccountDataAccess/src/Patient/PatientRepository.cs:149]: Enum.GetName could be simplified here with ds.ToString() or our EnumUtility.ToEnumString(ds).

**Pull request merged**